Client/Frontend/Home/ActivityStreamPanel.swift

```diff
@@ -0,0 +1,351 @@
+        return Section(section).headerView
+    }
+
+    func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
+        return Section(indexPath.section).cellHeight(self.traitCollection)
+    }
+
+    func showSiteWithURL(url: NSURL) {
+        let visitType = VisitType.Bookmark
+        homePanelDelegate?.homePanel(self, didSelectURL: url, visitType: visitType)
+    }
+
+    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
+        switch Section(indexPath.section) {
+        case .History:
+            let site = self.history[indexPath.row]
+            showSiteWithURL(NSURL(string:site.url)!)
+        case .TopSites:
+            return
+        } 
+    }
+
+}
+
+// MARK: - Tableview Data Source
+extension ActivityStreamPanel: UITableViewDataSource {
+
+    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
+        return Section.count
+    }
+
+    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        switch Section(section) {
+            case .TopSites:
+                return topSiteHandler.content.isEmpty ? 0 : 1
+            case .History:
+                 return self.history.count
+        }
+    }
+
+    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
+        let identifier = Section(indexPath.section).cellIdentifier
+        let cell = tableView.dequeueReusableCellWithIdentifier(identifier, forIndexPath: indexPath)
+
+        switch Section(indexPath.section) {
+        case .TopSites:
+            return configureTopSitesCell(cell, forIndexPath: indexPath)
+        case .History:
+            return configureHistoryItemCell(cell, forIndexPath: indexPath)
+        }
+    }
+
+    func configureTopSitesCell(cell: UITableViewCell, forIndexPath indexPath: NSIndexPath) -> UITableViewCell {
+        let topSiteCell = cell as! ASHorizontalScrollCell
+        topSiteCell.delegate = self.topSiteHandler
+        return cell
+    }
+
+    func configureHistoryItemCell(cell: UITableViewCell, forIndexPath indexPath: NSIndexPath) -> UITableViewCell {
+        let simpleHighlightCell = cell as! SimpleHighlightCell
+        let site = history[indexPath.row]
+        simpleHighlightCell.configureSimpleHighlightCell(site)
+        return simpleHighlightCell
+    }
+
+}
+
+// MARK: - Data Management
+extension ActivityStreamPanel {
+
+    func notificationReceived(notification: NSNotification) {
+        switch notification.name {
+        case NotificationProfileDidFinishSyncing:
+            // Only reload top sites if there the cache is dirty since the finish syncing
+            // notification is fired everytime the user re-enters the app from the background.
+            self.profile.history.areTopSitesDirty(withLimit: ASPanelUX.topSitesCacheSize) >>== { dirty in
```

Comment:
We already have the `dirty` check within `invalidateTopSites`. Does it make sense to have this here?